prov/rxd/src/rxd_cq.c

```diff
@@ -229,391 +231,519 @@ static void rxd_ep_recv_data(struct rxd_ep *ep, struct rxd_x_entry *rx_entry,
 		err_entry.prov_errno = 0;
 		rxd_cq_report_error(rx_cq, &err_entry);
 	}
-	fastlock_release(&ep->util_ep.rx_cq->cq_lock);
 
+	ep->peers[rx_entry->peer].rx_msg_id = rx_entry->msg_id;
 	rxd_rx_entry_free(ep, rx_entry);
 }
 
-static int rxd_check_pkt_ids(struct rxd_x_entry *rx_entry,
-			     struct rxd_pkt_hdr pkt_hdr)
+static int rxd_comp_pkt_msg_id(struct dlist_entry *item, const void *arg)
+{
+	struct rxd_op_pkt *list_op = (struct rxd_op_pkt *)
+				     ((container_of(item,
+				     struct rxd_pkt_entry, d_entry))->pkt);
+	struct rxd_op_pkt *new_op = (struct rxd_op_pkt *)
+				    ((container_of((struct dlist_entry *) arg,
+				    struct rxd_pkt_entry, d_entry))->pkt);
+
+	return (new_op->pkt_hdr.msg_id > list_op->pkt_hdr.msg_id) ? 1 : 0;
+}
+
+static int rxd_comp_rx_msg_id(struct dlist_entry *item, const void *arg)
+{
+	struct rxd_x_entry *list_rx = container_of(item, struct rxd_x_entry, entry);
+	struct rxd_x_entry *new_rx = container_of((struct dlist_entry *) arg,
+						  struct rxd_x_entry, entry);
+
+	return (new_rx->msg_id > list_rx->msg_id) ? 1 : 0;
+}
+
+static void rxd_progress_buf_cq(struct rxd_ep *ep, fi_addr_t peer)
+{
+	struct rxd_x_entry *rx_entry;
+
+	dlist_foreach_container(&ep->peers[peer].buf_cq, struct rxd_x_entry,
+				rx_entry, entry) {
+		if (rx_entry->msg_id != ep->peers[peer].rx_msg_id + 1)
+			return;
+
+		rxd_complete_rx(ep, rx_entry);
+	}
+}
+
+static void rxd_ep_recv_data(struct rxd_ep *ep, struct rxd_x_entry *rx_entry,
+			     struct rxd_data_pkt *pkt, size_t size)
+{
+	struct rxd_domain *rxd_domain = rxd_ep_domain(ep);
+	uint64_t done;
+
+	done = ofi_copy_to_iov(rx_entry->iov, rx_entry->iov_count,
+			       rxd_domain->max_inline_sz +
+			       (pkt->pkt_hdr.seg_no * rxd_domain->max_seg_sz),
+			       pkt->msg, size - RXD_DATA_HDR_SIZE - ep->prefix_size);
+
+	rx_entry->bytes_done += done;
+	ep->peers[pkt->pkt_hdr.peer].rx_seq_no++;
+	rx_entry->next_seg_no++;
+
+	if (rx_entry->next_seg_no < rx_entry->num_segs) {
+		if (!(ep->peers[pkt->pkt_hdr.peer].rx_seq_no %
+		    ep->peers[pkt->pkt_hdr.peer].rx_window))
+			rxd_ep_send_ack(ep, pkt->pkt_hdr.peer);
+		return;
+	}
+	rxd_ep_send_ack(ep, pkt->pkt_hdr.peer);
```

Comment:
Possible future enhancement: if we move up sending the ack, it may provide some overlap with ofi_copy_to_iov.